src/bin/tikv-ctl.rs

```diff
@@ -21,25 +21,415 @@ extern crate clap;
+                        }
+                    }
+                }
+                if cfs.contains(&CF_WRITE) {
+                    for write_info in mvcc.take_writes().into_iter() {
+                        if start_ts.map_or(true, |ts| write_info.get_start_ts() == ts) &&
+                            commit_ts.map_or(true, |ts| write_info.get_commit_ts() == ts)
+                        {
+                            // FIXME: short_value is lost in kvproto.
+                            println!("\t write cf value: {}", print_to_string(&write_info));
+                        }
+                    }
+                }
+                println!("");
+                future::ok::<(), String>(())
+            },
+        );
+        if let Err(e) = scan_future.wait() {
+            eprintln!("{}", e);
+            process::exit(-1);
+        }
+    }
+
+    fn diff_region(
+        &self,
+        region: u64,
+        db: Option<&str>,
+        raft_db: Option<&str>,
+        host: Option<&str>,
+    ) {
+        let rhs_debug_executor = new_debug_executor(db, raft_db, host);
+
+        let r1 = self.get_region_info(region);
+        let r2 = rhs_debug_executor.get_region_info(region);
+        println!("region id: {}", region);
+        println!("db1 region state: {:?}", r1.region_local_state);
+        println!("db2 region state: {:?}", r2.region_local_state);
+        println!("db1 apply state: {:?}", r1.raft_apply_state);
+        println!("db2 apply state: {:?}", r2.raft_apply_state);
+
+        match (r1.region_local_state, r2.region_local_state) {
+            (None, None) => return,
+            (Some(_), None) | (None, Some(_)) => {
+                println!("db1 and db2 don't have same region local_state");
+                return;
+            }
+            (Some(region_local_1), Some(region_local_2)) => {
+                let start_key = keys::data_key(region_local_1.get_region().get_start_key());
+                let end_key = keys::data_key(region_local_1.get_region().get_end_key());
+                if start_key != keys::data_key(region_local_2.get_region().get_start_key()) ||
+                    end_key != keys::data_key(region_local_2.get_region().get_end_key())
+                {
+                    println!("db1' range doesn't equal to db2' range.");
+                    return;
+                }
+                let mvcc_infos_1 = self.get_mvcc_infos(start_key.clone(), end_key.clone(), 0);
+                let mvcc_infos_2 = rhs_debug_executor.get_mvcc_infos(start_key, end_key, 0);
+                let cmp_future = mvcc_infos_1.zip(mvcc_infos_2).for_each(|(item_1, item_2)| {
+                    let (key1, mvcc1) = (escape(&item_1.0), item_1.1);
+                    let (key2, mvcc2) = (escape(&item_2.0), item_2.1);
+                    if key1 != key2 {
+                        let err_msg = format!("db1 cur key: {}, db2 cur key: {}", key1, key2);
```

Comment:
Is it OK to return early here? /cc @zhangjinpeng1987

Reply:
Do you mean we return a `future::err` here but not a `future::ok`, which causes the steping on the `DBIterator`s stop? I think it's OK if the future lib can deconstruct those `DBIterator`s correctly.

Reply:
We need to print all different keys.